### PR TITLE
Propagate never-returning functions, except when annotated with [@opaque]

### PR DESCRIPTION
**Pull request by Ekdohibs**
When a function never returns, this information is propagated and used, even if the function is not inlined. This behaviour can be disabled by an `[@opaque]` attribute on the function, which automatically implies `[@inline never]` and `[@local never]` as well.

### CODE REVIEW

File: ocaml/testsuite/tests/backtrace/pr6920_why_at.ml [hunk 1]
```diff
@@ -5,7 +5,7 @@
    exit_status = "2"
 *)
 
-let why : unit -> unit = fun () -> raise Exit [@@inline never]
+let why : unit -> unit = fun () -> raise Exit [@@inline never][@@opaque]
```
Comment (lthls) [inline ocaml/testsuite/tests/backtrace/pr6920_why_at.ml:8]:
```suggestion
let why : unit -> unit = fun () -> raise Exit [@@opaque]
```

File: ocaml/testsuite/tests/backtrace/pr6920_why_swallow.ml [hunk 2]
```diff
@@ -5,7 +5,7 @@
    exit_status = "2"
 *)
 
-let why : unit -> unit = fun () -> raise Exit [@@inline never]
+let why : unit -> unit = fun () -> raise Exit [@@inline never][@@opaque]
```
Comment (lthls) [inline ocaml/testsuite/tests/backtrace/pr6920_why_swallow.ml:8]:
```suggestion
let why : unit -> unit = fun () -> raise Exit [@@opaque]
```

**Review by lthls — APPROVED** [review]
Looks good.
I've left suggestions that I think would make the tests a bit cleaner (if they work as I intend them to).

**Comment by mshinwell** [comment]
I haven't read the details, but LGTM in principle.  OK to merge once the suggestions addressed.

**Pull request merged**